polars/polars-core/src/frame/asof_join/groups.rs

```diff
@@ -730,4 +801,143 @@ mod test {
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_asof_by4() -> PolarsResult<()> {
+        let trades = df![
+            "time" =>    [23i64,     38,     48,     48,     48],
+            "ticker" => ["MSFT", "MSFT", "GOOG", "GOOG", "AAPL"],
+            "groups_numeric" => [1, 1, 2, 2, 3],
+            "bid" => [51.95, 51.95, 720.77, 720.92, 98.0]
+        ]?;
+
+        let quotes = df![
+            "time" =>    [23i64,     23,     30,     41,     48,     49,     72,     75],
+            "ticker" => ["GOOG", "MSFT", "MSFT", "MSFT", "GOOG", "AAPL", "GOOG", "MSFT"],
+            "bid" =>     [720.5,  51.95,  51.97,  51.99,  720.5,  97.99,  720.5,  52.01],
+            "groups_numeric" => [2, 1, 1, 1, 2, 3, 2, 1],
+
+        ]?;
+        /*
+        trades:
+        shape: (5, 4)
+        ┌──────┬────────┬────────────────┬────────┐
+        │ time ┆ ticker ┆ groups_numeric ┆ bid    │
+        │ ---  ┆ ---    ┆ ---            ┆ ---    │
+        │ i64  ┆ str    ┆ i32            ┆ f64    │
+        ╞══════╪════════╪════════════════╪════════╡
+        │ 23   ┆ MSFT   ┆ 1              ┆ 51.95  │
+        ├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
+        │ 38   ┆ MSFT   ┆ 1              ┆ 51.95  │
+        ├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
+        │ 48   ┆ GOOG   ┆ 2              ┆ 720.77 │
+        ├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
+        │ 48   ┆ GOOG   ┆ 2              ┆ 720.92 │
+        ├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
+        │ 48   ┆ AAPL   ┆ 3              ┆ 98.0   │
+        └──────┴────────┴────────────────┴────────┘
+        quotes:
+        shape: (8, 4)
+        ┌──────┬────────┬───────┬────────────────┐
+        │ time ┆ ticker ┆ bid   ┆ groups_numeric │
+        │ ---  ┆ ---    ┆ ---   ┆ ---            │
+        │ i64  ┆ str    ┆ f64   ┆ i32            │
+        ╞══════╪════════╪═══════╪════════════════╡
+        │ 23   ┆ GOOG   ┆ 720.5 ┆ 2              │
+        ├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ 23   ┆ MSFT   ┆ 51.95 ┆ 1              │
+        ├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ 30   ┆ MSFT   ┆ 51.97 ┆ 1              │
+        ├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ 41   ┆ MSFT   ┆ 51.99 ┆ 1              │
+        ├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ 48   ┆ GOOG   ┆ 720.5 ┆ 2              │
+        ├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ 49   ┆ AAPL   ┆ 97.99 ┆ 3              │
+        ├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ 72   ┆ GOOG   ┆ 720.5 ┆ 2              │
+        ├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ 75   ┆ MSFT   ┆ 52.01 ┆ 1              │
+        └──────┴────────┴───────┴────────────────┘
+        */
+
+        let out = trades.join_asof_by(
+            &quotes,
+            "time",
+            "time",
+            ["ticker"],
+            ["ticker"],
+            AsofStrategy::Forward,
+            None,
+        )?;
```

Comment:
I'm not sure why you'd ever be looking for quotes from after the trades, but who said that test cases have to make sense in terms of business logic? 🤷

Reply:
I don't understand financial business, so to me it seems like a perfectly valid test. ;)